src/Orleans/Properties/AssemblyInfo.cs

```diff
@@ -54,4 +54,5 @@ MIT License
 [assembly: InternalsVisibleTo("UnitTestGrainInterfaces, PublicKey=0024000004800000940000000602000000240000525341310004000001000100eb81539dc2728946d4860bad0fbc91ea23d7a835e00547aa4e9a7abf185535b564d29dbe77915b779951fef2f4b1ab39cc3a246cf572c7b217b5c892835a687eb22638183d66930039ff9fa6e7e71cb880f854fb76b884dfc905bf1f3c623abad7d4732c1ce62efff937155e352170831239acbe103a41ed363a90001d3108ca")]
 [assembly: InternalsVisibleTo("UnitTestGrains, PublicKey=0024000004800000940000000602000000240000525341310004000001000100eb81539dc2728946d4860bad0fbc91ea23d7a835e00547aa4e9a7abf185535b564d29dbe77915b779951fef2f4b1ab39cc3a246cf572c7b217b5c892835a687eb22638183d66930039ff9fa6e7e71cb880f854fb76b884dfc905bf1f3c623abad7d4732c1ce62efff937155e352170831239acbe103a41ed363a90001d3108ca")]
 [assembly: InternalsVisibleTo("TesterInternal, PublicKey=0024000004800000940000000602000000240000525341310004000001000100eb81539dc2728946d4860bad0fbc91ea23d7a835e00547aa4e9a7abf185535b564d29dbe77915b779951fef2f4b1ab39cc3a246cf572c7b217b5c892835a687eb22638183d66930039ff9fa6e7e71cb880f854fb76b884dfc905bf1f3c623abad7d4732c1ce62efff937155e352170831239acbe103a41ed363a90001d3108ca")]
+[assembly: InternalsVisibleTo("Tester, PublicKey=0024000004800000940000000602000000240000525341310004000001000100eb81539dc2728946d4860bad0fbc91ea23d7a835e00547aa4e9a7abf185535b564d29dbe77915b779951fef2f4b1ab39cc3a246cf572c7b217b5c892835a687eb22638183d66930039ff9fa6e7e71cb880f854fb76b884dfc905bf1f3c623abad7d4732c1ce62efff937155e352170831239acbe103a41ed363a90001d3108ca")]
```

Comment:
InternalsVisibleTo was needed because the system grains are internal classes. Thinking about how to make the DI implementation simpler AND support the registration of system classes/grains I have this proposal, what do you think?

Making the interface internal
```
    internal interface IDependencyResolverProvider : IProvider
    {
        IDependencyResolver GetDependencyResolver(ClusterConfiguration config, NodeConfiguration nodeConfig, TraceLogger logger);
    }
```

Provide a default implementation with 1 method to override:
```
    public abstract class DependencyResolverProviderBase : IDependencyResolverProvider
    {
        public string Name
        {
            get { return "Orleans DependencyResolver Provider"; }
        }

        public System.Threading.Tasks.Task Init(string name, IProviderRuntime providerRuntime, IProviderConfiguration config)
        {
            //
            // This method will not be called by the DependenceResolverProviderManager, when loading the provider.
            //

            throw new System.NotSupportedException();
        }

        public abstract IDependencyResolver ConfigureResolver(params System.Type[] systemTypesToRegister);

        public IDependencyResolver GetDependencyResolver(ClusterConfiguration config, NodeConfiguration nodeConfig, TraceLogger logger)
        {
            var dependencyResolver = ConfigureResolver(
                typeof(GrainBasedMembershipTable), 
                typeof(GrainBasedReminderTable), 
                typeof(GrainBasedPubSubRuntime));

            return dependencyResolver;
        }
    }
```

So a project specific implementation can be simple like this:
```
    public class AutofacDependencyResolverProvider : DependencyResolverProviderBase
    {
        public virtual IDependencyResolver ConfigureResolver(System.Type[] systemTypesToRegister)
        {
            var builder = new ContainerBuilder();

            builder.RegisterTypes(systemTypesToRegister);

            //
            // We've to register the concrete type and the interface of the grain too:
            // - the test code is retrieving grains based on the interface.
            // - Orleans messaging is retrieving grains by their type.
            //
            // DI container grain registration helper functions can be written to make grain registration
            // a no brainer. One thing to note: at this point the type manager did not load all the assemblies,
            // so to support grain registration based on the loaded assemblies assembly loader should be invoked
            // earlier in Silo startup.
            //

            builder.RegisterType<SimpleGrain>();
            builder.RegisterType<SimpleGrain>().AsImplementedInterfaces<ISimpleGrain, ConcreteReflectionActivatorData>();

            var container = builder.Build();

            return new AutofacOrleansDependencyResolver(container);
        }
}
```

This way, we've need to add InternalsVisibleTo for Orleans.Providers project, which is ok, I think.